### DISCUSSION TITLE
Javascript issues with Rhino and Java 17 when calling Java code: IllegalReflectiveAccess

### DISCUSSION THREAD

**Discussion by hvbtup** [eclipse-birt/birt#1138] [Q&A]
There are issues when calling Java libraries from Javascript with Java 17.

With Java 17, it fails with an exception similar to this:
java.lang.IllegalAccessException: class org.mozilla.javascript.MemberBox cannot access class sun.java2d.SunGraphics2D (in module java.desktop) because module java.desktop does not export sun.java2d to unnamed module

With Java 11, it works, but results in a warning:
WARNING: Illegal reflective access by org.mozilla.javascript.MemberBox (...) to method sun.java2d.SunGraphics2D.setColor(java.awt.Color)

This is described here:

https://groups.google.com/g/mozilla-rhino/c/jmNIszUpBAE
https://github.com/ckeditor/ckbuilder/issues/34
https://github.com/mozilla/rhino/issues/462
https://github.com/mozilla/rhino/issues/1269

The last one seems to be the most up-to-date.

Note: This seems to be an issue with Rhino and Java's module system.

A possible workaround is to use the following VM argument:

    --add-exports java.desktop/sun.java2d=ALL-UNNAMED

But I think this can only be a short-term workaround.

In our environment, we stumble upon this issue, because in many of our reports, we are accessing a Java barcode library (to be more precise: from IDAutomation) through Javascript. This library needs a `java.awt.BufferedImage` to render the barcode. We want to prepare that image by filling it all white. To do this, we call the `setColor` and `fillRect` methods.
These methods fail with Java 17.

A code excerpt of our barcode JS function:

        var bc = new com.idautomation.datamatrix.DataMatrix();
        ...
        // create image
        var t = java.awt.image.BufferedImage.TYPE_BYTE_INDEXED;
        var siz = bc.getSize();
        var image = new java.awt.image.BufferedImage(siz.width, siz.height, t);

        // get graphic context of image
        var imgGraphics = image.createGraphics();

        imgGraphics.setColor(java.awt.Color.WHITE);
        imgGraphics.fillRect(0, 0, siz.width, siz.height);

        // paint DataMatrix in graphics context of image
        bc.paint(imgGraphics);
        ...
        // get the image's bytes in PNG format
        var baos = new java.io.ByteArrayOutputStream();
        javax.imageio.ImageIO.write(image, "png", baos);
        var bytes = baos.toByteArray();

        return bytes;

You can test this yourself by commenting the lines with `bc` - the funtion will then just a white rect as PNG bytes, you can display it as a dynamic image.

**Comment by hvbtup** [comment]
Since this is a Rhino issue, this is related to #732

**Comment by claesrosell** [comment]
Can you provide a complete stacktrace for this?
I suspect that the error is not with Rhino.

**Reply by hvbtup** [reply]
See this tiny example report and the stack trace.
[red-rectangle.zip](https://github.com/eclipse/birt/files/10267709/red-rectangle.zip)

I'm pretty sure this *is* a Rhino issue.

**Reply by speckyspooky** [reply]
I would confirm that this is an rhino issue because we have since Java 9 the new JVM module-handling
and with Java 16 some default jars was extracted from JVM to special modules and these classes are not accessible on rhino 1.7.10 by default.
I had an similar issue with BIRT 4.12 and JVM 17 - #1160.

It seems that the newest version 1.17.14 have some fixes on it. But it could be that not all necessary fixes are done - comments of october https://github.com/mozilla/rhino/issues/1269.
